Add initial libstd support for Xous

xobs opened the pull request:
This patchset adds some minimal support to the tier-3 target `riscv32imac-unknown-xous-elf`. The following features are supported:

* alloc
* thread creation and joining
* thread sleeping
* thread_local
* panic_abort
* mutex
* condvar
* stdout

Additionally, internal support for the various Xous primitives surrounding IPC have been added as part of the Xous FFI. These may be exposed as part of `std::os::xous::ffi` in the future, however for now they are not public.

This represents the minimum viable product. A future patchset will add support for networking and filesystem support.

xobs committed:
libstd: add xous to libstd

Add the `xous` target to libstd. Currently this defers everything to the
`unsupported` target.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add os-specific ffi calls

Xous has no C FFI. Instead, all FFI is done via syscalls that are
specified in Rust. Add these FFI calls to libstd, as well as some of the
currently-supported syscalls.

This enables Rust programs to interact with the Xous operating system
while avoiding adding an extra dependency to libstd.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: the basics of `os`

Add the basics to get the operating system running, including how to
exit the operating system.

Since Xous has no libc, there is no default entrypoint. Add a `_start`
entrypoint to the system-specific os module.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add services support

Xous has a concept of `services` that provide various features.
Processes may connect to these services by name or by address. Most
services require a name server in order to connect.

Add a file with the most common services, and provide a way to connect
to a service by querying the name server.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add output support for stdio

Add support for stdout. This enables basic console printing via
`println!()`. Output is written to the log server.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add thread support

Add initial support for threads on Xous. This includes thread creation
and joining.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add support for locks

Add support for Condvar, Mutex, and RWLock. These are all backed by the
ticktimer server.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add thread_local_key

Add an implementation of thread local storage. This uses a container
that is pointed to by the otherwise-unsed `$tp` register. This container
is allocated on-demand, so threads that use no TLS will not allocate
this extra memory.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
panic_abort: call __rust_abort on xous

Xous does not properly handle invalid instructions. Instead, call
`__rust_abort` and exit normally.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
Please let me know if there is anything I should do in order to make this merge easier, please let me know. This is my first relatively major change to libstd.

Note that this will not build unless you have the latest `llvm-project` which includes https://github.com/llvm/llvm-project/commit/7c80e7a2943ade7b999c14ad582b52035098e5ae but it does not require any external dependencies beyond a C compiler capable of building RISC-V code.

Nilstrieb commented on library/std/src/sys/xous/locks/rwlock.rs:
```diff
@@ -0,0 +1,67 @@
+use crate::cell::Cell;
+
+pub struct RwLock {
+    // This platform has no threads, so we can use a Cell here.
+    mode: Cell<isize>,
```
You've implemented threads, so this is is probably a mistake

xobs replied:
That's a very good point, thank you. I've replaced it with a version that uses `AtomicISize` and now spins while it waits for locks, calling `do_yield()` in between.

Ideally we'd delegate this to an external server to coordinate in case of contention, but I think that can wait for now.

bors commented:
:umbrella: The latest upstream changes (presumably #103723) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fbstj commented on library/std/src/sys/xous/time.rs:
```diff
@@ -0,0 +1,59 @@
+    pub fn now() -> Instant {
+        let result = crate::os::xous::ffi::blocking_scalar(
+            crate::os::xous::services::ticktimer_server(),
+            [0 /* ElapsedMs */, 0, 0, 0, 0],
```
is it worth making all the `ElapsedMs`-like things into an enum or named-consts somewhere?

or structs which have an `Into<[usize; 5]>` (or a different trait used as the `*_scalar()` argument) and appropriate constructors that wrap/document the other possible arguments?

xobs replied:
I do like that design idea. I initially made the `ffi` functions all `pub`, because it would be really nice to provide that functionality to all users. For example, I'd like to add it to `getrandom` in order to talk to the TRNG server. However, I made them `pub(crate)` in order to allow for suggestions such as this one.

I think the API arguments probably should go under `crate::os::xous::services`.

xobs replied:
The latest push uses this approach.

xobs commented:
I rebased to the latest `master` branch and incorporated the suggestions that were recommended.

albertlarsan68 commented:
This does not affect the bootstrap code, do @rustbot label -A-bootstrap.

xobs committed:
std: net: skip tests on xous

Network functionality is not yet ready for merging.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: fs: skip fs tests on xous

The xous filesystem support is not yet ready for merging.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: process: skip tests on xous

Xous does not yet support spawning processes.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
I'm in the process of reworking this so that it passes tests. I've gotten libunwind functioning, and now I can actually run the Rust libstd test suite.

I've discovered some issues in the TLS implementation, as well as potential issues in the Condvar or Mutex implementation. I will work on those issues and remove the `Draft` status once they are addressed.

bors commented:
:umbrella: The latest upstream changes (presumably #107405) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@xobs any updates on this?

xobs committed:
std: xous: add alloc support

Basic alloc support on Xous is supported by the `dlmalloc` crate. This
necessitates bumping the dlmalloc version to 0.2.4.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
bootstrap: builder: don't add origin paths on xous

Don't add the origin rpath when calling the linker. This is unnecessary
on Xous since there is no dynamic linker, and Xous can use an ordinary
bare-metal linker with no rpath options.

As part of this patch, the logic was inverted so that the "else" clause
contains everything except "windows" and "xous".

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
std: xous: add support for time

Add support for determining the current time. This connects to the
ticktimer server in order to get the system uptime.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
Thank you for the nudge. The dependency tree is fairly large, so I will split it into multiple PRs.

I have reworked it based on some of my outstanding patches, and the big piece missing is support for unwinding. After chatting with people on Zulip, I'll work to get `unwinding` working rather than submitting patches to rust's llvm branch in order to get libunwind working.

bjorn3 commented on src/bootstrap/lib.rs:
```diff
@@ -754,6 +754,9 @@ impl Build {
             LlvmLibunwind::No => {}
         }
+        if target.contains("xous") {
+            features.push_str(" compiler-builtins-mem");
```
You could enable this directly in compiler-builtin's build script. That way `cargo build -Zbuild-std` will also get this enabled automatically.

xobs replied:
That's a really good suggestion. I've created https://github.com/rust-lang/compiler-builtins/pull/548 that does this, and I've tested that it works locally.

xobs replied:
With 0.1.101 released, I've removed the patch and updated `Cargo.lock` to pull in the latest version.

xobs committed:
Cargo.lock: bump compiler_builtins to 0.1.101

The Xous operating system has no libc, and relies on `compiler_builtins`
to supply basic functions such as `memcpy`. Bump the version to 0.1.101
to pull in a version of this crate with the flag enabled for Xous.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
As suggested by @bjorn3 , I removed the explicit addition of `compiler-builtins-mem` and instead added it to `compiler_builtins`. This was released as part of `0.1.101`. I then rebased on top of `master` and added a patch to bump the version of `compiler_builtins`.

bjorn3 commented on library/std/src/sys/xous/time.rs:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    pub fn checked_add_duration(&self, other: &Duration) -> Option<Instant> {
+        Some(Instant(self.0.checked_add(*other)?))
```
```suggestion
        self.0.checked_add(*other).map(Instant)
```
should work too I think. I don't have strong feelings either way.

bjorn3 commented on library/std/src/sys/xous/thread_local_key.rs:
```diff
@@ -0,0 +1,184 @@
+///
+/// The index into this register is the `key`. This key is identical
+/// between all threads, but indexes a different offset within this
+/// pointer.
```
Could you either remove the empty line between the doc comments and the item it applies to or change it into a module doc comment?

bjorn3 commented on library/std/src/sys/xous/thread.rs:
```diff
@@ -0,0 +1,141 @@
+
+        // No access to this page. Note: Write-only pages are illegal, and will
+        // cause an access violation.
+        crate::os::xous::ffi::update_memory_flags(
```
```suggestion
        update_memory_flags(
```

bjorn3 commented on library/std/src/sys/xous/thread.rs:
```diff
@@ -0,0 +1,141 @@
+        let stack_plus_guard_pages: &mut [u8] = map_memory(
+            None,
+            None,
+            stack_size + GUARD_PAGE_SIZE + GUARD_PAGE_SIZE,
```
Maybe
```suggestion
            GUARD_PAGE_SIZE + stack_size + GUARD_PAGE_SIZE,
```
to indicate that the stack is sandwiched between guard pages?

xobs replied:
Thank you for spotting this! I intended to write it this way, and that's why I wrote it that way further down.

bjorn3 commented on library/std/src/sys/xous/thread.rs:
```diff
@@ -0,0 +1,141 @@
+
+        let tid = create_thread(
+            thread_start as *mut usize,
+            &stack_plus_guard_pages[GUARD_PAGE_SIZE..(stack_size - GUARD_PAGE_SIZE)],
```
```suggestion
            &stack_plus_guard_pages[GUARD_PAGE_SIZE..(stack_size + GUARD_PAGE_SIZE)],
```
right? If `stack_size` is `MIN_STACK_SIZE`, `stack_size - GUARD_PAGE_SIZE` would be 0.

xobs replied:
Another good catch. I must have been thinking it was `[start..length]` rather than `[start..end]`.

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,631 @@
+}
+
+/// Destroy the given memory, returning it to the compiler.
+pub(crate) fn unmap_memory<T>(range: &mut [T]) -> Result<(), Error> {
```
I don't think this is right. `&mut [T]` as argument will assert that the slice stays alive until the end of the function, which it clearly does not. `*mut [T]` and an unsafe function probably makes more sense.

xobs replied:
That's an interesting approach. I'd never considered `*mut [T]` before. It has the nice property of indicating how unsafe the function call is while still providing a fat pointer.

bjorn3 commented on library/std/src/sys/xous/thread.rs:
```diff
@@ -0,0 +1,141 @@
+            unsafe {
+                asm!("ret", in("a0") 0, in("ra") 0xff80_3000usize,
+                    options(nomem, nostack, noreturn)
+                );
```
This asm block will need to be merged with the previous one. asm blocks can be implemented as function calls (eg my cranelift based backend does this because cranelift doesn't have native inline asm support), which doesn't work if the stack is deallocated.

bjorn3 commented on library/std/src/sys/xous/thread.rs:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            // Exit the thread by returning to the magic address 0xff80_3000usize,
+            // which tells the kernel to deallcate this thread.
```
```suggestion
            // which tells the kernel to deallocate this thread.
```
nit

bjorn3 commented on library/std/src/sys/xous/stdio.rs:
```diff
@@ -0,0 +1,131 @@
+            for (dest, src) in lend_buffer.0.iter_mut().zip(chunk) {
+                *dest = *src;
+            }
+            crate::os::xous::ffi::lend(connection, 1, &lend_buffer.0, 0, chunk.len()).unwrap();
```
```suggestion
            lend(connection, 1, &lend_buffer.0, 0, chunk.len()).unwrap();
```

bjorn3 commented on library/std/src/sys/xous/stdio.rs:
```diff
@@ -0,0 +1,131 @@
+
+impl io::Read for Stdin {
+    fn read(&mut self, _buf: &mut [u8]) -> io::Result<usize> {
+        Ok(0)
```
Xous has no stdin?

xobs replied:
Not at this time. It will someday, but for now it's just a stub. Its primary interface is through a graphical chat program. It may make sense to add stdin in the future as something that's hooked up to some GUI library, but right now all of this stdout is going to a logging server which is doing the printing.

bjorn3 commented on library/std/src/sys/xous/os.rs:
```diff
@@ -0,0 +1,130 @@
+}
+
+pub fn error_string(errno: i32) -> String {
+    core::convert::Into::<crate::os::xous::ffi::Error>::into(errno).to_string()
```
`Into` should be in the prelude, right?

xobs replied:
Good point. I copied that line directly from the compiler, and error messages don't seem to pay attention to the prelude. I've adjusted it to remove the scoping.

bjorn3 commented on library/std/src/sys/xous/locks/condvar.rs:
```diff
@@ -0,0 +1,119 @@
+        //         self.index(),
+        //         *counter
+        //     );
+        // }
```
Stray comment

bjorn3 commented on library/std/src/sys/xous/alloc.rs:
```diff
@@ -0,0 +1,62 @@
+unsafe impl GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        // SAFETY: DLMALLOC access is guranteed to be safe because the lock gives us unique and non-reentrant access.
```
*guaranteed

bjorn3 commented on library/std/src/os/xous/services/log.rs:
```diff
@@ -0,0 +1,63 @@
+// Group `usize` bytes into a `usize` and return it, beginning
+// from `offset` * sizeof(usize) bytes from the start. For example,
+// `group_or_null([1,2,3,4,5,6,7,8], 1)` on a 32-bit system will
+// return a usize with 5678 packed into it.
```
Maybe turn this into a doc comment?

xobs replied:
Good idea

bjorn3 commented on library/std/src/os/xous/mod.rs:
```diff
@@ -0,0 +1,17 @@
+#![stable(feature = "rust1", since = "1.0.0")]
```
Should this be insta-stable?

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,631 @@
+/// Allocate memory from the system. An optional physical and/or virtual address
+/// may be specified in order to ensure memory is allocated at specific offsets,
+/// otherwise the kernel will select an address.
+pub(crate) fn map_memory<T>(
```
Can this ever overwrite an existing mapping? If so it will need to be unsafe.

xobs replied:
No, but it's possible to specify a virtual address, and in doing so it's possible to effectively clone pointers.

I'll mark it as `unsafe` and note why.

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,631 @@
+/// Adjust the memory flags for the given range. This can be used to remove flags
+/// from a given region in order to harden memory access. Note that flags may
+/// only be removed and may never be added.
+pub(crate) fn update_memory_flags<T>(range: &mut [T], new_flags: MemoryFlags) -> Result<(), Error> {
```
`&mut [T]` would be wrong if the memory range ends up immutable.

xobs replied:
My thinking here was that the handle was mutable but the actual data was not. Also, `&mut` indicates the caller has exclusive access to the memory, preventing anyone else from accessing it. Especially since it's not possible to hand slices around, so I can't have the function take ownership of the memory.

Do you have another suggestion? Or should I simply abandon the idea of passing allocated memory around as `&[u8]`? Or would a newtype be better here, since that can be passed around?

bjorn3 replied:
Using `*mut [T]` with an unsafe function would be better I think.

xobs replied:
I've changed this to `*mut [T]` and have marked the function as `unsafe`. Additionally, I have added a note to the function describing safety.

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,631 @@
+/// Create a thread with a given stack and up to four arguments
+pub(crate) fn create_thread(
+    start: *mut usize,
+    stack: &[u8],
```
`&[u8]` doesn't allow mutation of the stack.

xobs committed:
std: enable feature(slice_ptr_len) on xous

Xous passes slice pointers around in order to manipulate memory.
This is feature-gated behind `slice_ptr_len`. Xous is currently
the only target to use this feature, so gate it behind an OS flag.

Signed-off-by: Sean Cross <sean@xobs.io>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5876178742/job/15933819976) [(plain)](https://github.com/rust-lang/rust/commit/319ea3f6438f8bcc0f1a7b50f724f088ce1df810/checks/15933819976/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'actions/upload-artifact@v3' (SHA:0b7f8abb1508181956e8e162db84b466c27e18ce)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container d41b26b84e91
 ---> 36ec1ae69a56
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 66e1af799212
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180116 sha256=351235b2326fb4db7a18e257e13ce7896c5f77339521e2c2612e71e154800a19
  Stored in directory: /tmp/pip-ephem-wheel-cache-ztf6bo9l/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.3-py3-none-any.whl (3.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 15.1 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 25.4 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.10.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.2-py3-none-any.whl (10 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.2 platformdirs-3.10.0 virtualenv-20.24.3
Removing intermediate container 97f972cdbd19
 ---> a26f7b1b922c
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> ebc2f0011dc5
 ---> ebc2f0011dc5
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> a3b349b8966f
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 49b019870fd7
 ---> b5aa48b86692
Successfully built b5aa48b86692
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:b5aa48b866928a133dbf38f066132554bc2324d3b3162b28a3ed141c6c8844f5
Uploading finished image to https://ci-caches.rust-lang.org/docker/f76604104d6a6f4aa068dca8679411c51198f6c5562033174b89cf3ceb2f9ecc175231f1257b18e00fa57bb3eed5f53dbcc121e55de0452dda3db18894e27ea0

<botocore.awsrequest.AWSRequest object at 0x7f274e2ade10>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
DirectMap4k:      190400 kB
DirectMap2M:     6100992 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/4e3ce0e782a30d467ddbc542c3eb5d215eac1b9d/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-4e3ce0e782a30d467ddbc542c3eb5d215eac1b9d-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 26.85s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/lib.rs at line 624:
 #[stable(feature = "rust1", since = "1.0.0")]
 #[allow(deprecated, deprecated_in_future)]
 pub use core::{
-    assert_eq, assert_ne, debug_assert, debug_assert_eq, debug_assert_ne, matches, r#try, todo,
+    assert_eq, assert_ne, debug_assert, debug_assert_eq, debug_assert_ne, matches, todo, r#try,
     unimplemented, unreachable, write, writeln,
 
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/lib.rs" "/checkout/library/std/src/fs/tests.rs" "/checkout/library/std/src/backtrace/tests.rs" "/checkout/library/std/src/alloc.rs" "/checkout/library/std/src/f64/tests.rs" "/checkout/library/std/src/time.rs" "/checkout/library/std/build.rs" "/checkout/library/std/src/sys/itron/time.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Aug 16 08:04:41 UTC 2023
  network time: Wed, 16 Aug 2023 08:04:42 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

xobs commented:
I've made the adjustments you suggested. There are a few outstanding questions:

1. What is the correct  signature to `update_memory_flags()`? Ideally the argument would only have exclusive (i.e. Mutable) access to the range. Perhaps a newtype really is the best way here...
2. What is the correct way to express stability? We currently ship our own version libstd that builds against stable, so it would be nice to be able to ship a libstd based on upstream. Given that we're not a tier 2 target, and that there aren't any builders for RISC-V, I don't know what the best approach is.
3. I added a `#![config_attr(target_os = "xous", feature(slice_ptr_len))]` fate to `lib.rs`. Please let me know if this is the correct way to do it, or if I should change it somehow.

Thank you again for the review.

RalfJung commented on src/tools/miri/miri-script/src/commands.rs:
```diff
@@ -173,7 +173,7 @@ impl Command {
         // We pass `--no-verify` to avoid running git hooks like `./miri fmt` that could in turn
         // trigger auto-actions.
-        sh.write_file("rust-version", &commit)?;
+        sh.write_file("rust-version", format!("{commit}\n"))?;
```
Something went wrong here, I don't think this PR should change Miri. Looks like a rebase gone wrong, the diff shown here has already landed on master.

xobs replied:
I've rebased on top of master, but it looks like a change was made that broke `env`. Let me fix that and push a change.

xobs commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,637 @@
+/// Create a thread with a given stack and up to four arguments
+pub(crate) fn create_thread(
+    start: *mut usize,
+    stack: &mut [u8],
```
That's a good point. I've updated this to also be `*mut [u8]`.

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,637 @@
+
+    let result = a0;
+
+    if result == SyscallResult::MemoryReturned as usize {
```
If there is an error result does that mean the memory is no longer accessible by the caller?

xobs replied:
An error here means that the memory wasn't accessible already. For example, attempting to lend memory that you do not own, or attempting to lend memory that is not page-aligned.

If the memory is not page-aligned and page-sized, then they still have access to it.

If the memory wasn't owned by the caller, then it wasn't accessible to begin with, and nothing has changed. This can happen with a slice that was conjured out of thin air using `unsafe` functions. In theory it could happen if two threads had an immutable reference to the same slice, but in that case it would be behind a Mutex, and so such a circumstance would not be possible in safe Rust.

bjorn3 commented on library/std/src/os/xous/ffi.rs:
```diff
@@ -0,0 +1,637 @@
+    virt: Option<core::ptr::NonNull<T>>,
+    count: usize,
+    flags: MemoryFlags,
+) -> Result<&'static mut [T], Error> {
```
I think returning `&mut [T]` is fine, but I'm not entirely certain. @RalfJung?

RalfJung replied:
I lack all the context here, and it's a bit much to expect me to read a 2000 line diff to fill the gaps.^^ All the usual uniqueness guarantees must be upheld, of course.

Do you have a self-contained question about some concrete usage pattern of mutable references?

bjorn3 replied:
If `map_memory` returns `&'static mut [T]`, can `unmap_memory` unmap it again later through a raw pointer and `update_memory_flags` remove write permission for the memory region?

bjorn3 replied:
`unmap_memory` takes `*mut [T]` as argument and once https://github.com/rust-lang/rust/pull/104101#discussion_r1293787287 is fixed `update_memory_flags` would too.

RalfJung replied:
Sure, but I said "and transitive". A user taking `&mut [T]` and then calling `unmap_memory` on that would also be UB. That sounds pretty fragile to me, and it is why generally I would say it is a bad idea to use a `'static` reference for memory that will actually be deallocated.

xobs replied:
That's a good argument. I've updated it to take `*mut [T]`, and have added a note about safety. Additionally, the functions are now marked `unsafe`.

bjorn3 commented:
r=me with https://github.com/rust-lang/rust/pull/104101#discussion_r1293787287 fixed and confirmation that https://github.com/rust-lang/rust/pull/104101#discussion_r1301270685 and https://github.com/rust-lang/rust/pull/104101#discussion_r1301273290 are not unsound.

bjorn3 commented:
Could you reply to https://github.com/rust-lang/rust/pull/104101#discussion_r1301270685?

bjorn3 reviewed (approved):
LGTM, but @m-ou-se may want to take a look at it too. If not r=me.

xobs commented:
What are the next steps to getting this merged?

bjorn3 commented:
Given that @m-ou-se hasn't taken a look yet:

@bors r+

bors commented:
:pushpin: Commit 6c32a64962f591e6fd4cab72f63555b157684323 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 6c32a64962f591e6fd4cab72f63555b157684323 -->
<!-- homu: {"type":"Approved","sha":"6c32a64962f591e6fd4cab72f63555b157684323","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6c32a64962f591e6fd4cab72f63555b157684323 with merge ae9c330629f1fb03a636cb8af367fead024793db...
<!-- homu: {"type":"BuildStarted","head_sha":"6c32a64962f591e6fd4cab72f63555b157684323","merge_sha":"ae9c330629f1fb03a636cb8af367fead024793db"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6232578314/job/16919379660)
Approved by: bjorn3
Pushing ae9c330629f1fb03a636cb8af367fead024793db to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6232578314/job/16919379660"},"merge_sha":"ae9c330629f1fb03a636cb8af367fead024793db"} -->

rust-timer commented:
Finished benchmarking commit (ae9c330629f1fb03a636cb8af367fead024793db): [comparison URL](https://perf.rust-lang.org/compare.html?start=af78bae565e85b9c5698ee909af0652674eca6d4&end=ae9c330629f1fb03a636cb8af367fead024793db&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=af78bae565e85b9c5698ee909af0652674eca6d4&end=ae9c330629f1fb03a636cb8af367fead024793db&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=af78bae565e85b9c5698ee909af0652674eca6d4&end=ae9c330629f1fb03a636cb8af367fead024793db&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-4.1%, -1.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -4.4% | [-4.4%, -4.4%] | 1     |
| All ❌✅ (primary)                 | -2.6% | [-4.1%, -1.0%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=af78bae565e85b9c5698ee909af0652674eca6d4&end=ae9c330629f1fb03a636cb8af367fead024793db&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=af78bae565e85b9c5698ee909af0652674eca6d4&end=ae9c330629f1fb03a636cb8af367fead024793db&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.926s -> 631.537s (-0.06%)
**Artifact size**: 317.94 MiB -> 317.91 MiB (-0.01%)
